ibis/pandas/execution.py

```diff
@@ -0,0 +1,224 @@
+
+    predicates = op.predicates
+    if predicates:
+        data = data.loc[
+            reduce(operator.and_, (execute(p, scope) for p in predicates))
+        ]
+
+    gb = data.groupby([execute(by, scope) for by in op.by])
+
+    aggs = [reduction_to_lambda(agg_expr) for agg_expr in op.agg_exprs]
+    pieces = [func(gb) for func in aggs]
+    agg_df = pd.concat(pieces, axis=1)
+
+    having = op.having
+    if having:
+        agg_df = agg_df[
+            reduce(operator.and_, (execute(h, scope) for h in having))
+        ]
+
+    sort_keys = op.sort_keys
+    if sort_keys:
+        agg_df = agg_df.sort_values(sort_keys)
+    return agg_df.reset_index()
+
+
+@execute_node.register(ops.Reduction, pd.Series)
+def execute_reduction_series(op, data, scope=None):
+    return getattr(data, type(op).__name__.lower())()
+
+
+@execute_node.register(ops.Reduction, pd.Series, pd.Series)
+def execute_masked_reduction_series(op, data, mask, scope=None):
+    return getattr(data[mask], type(op).__name__.lower())()
+
+
+_JOIN_TYPES = {
+    ops.LeftJoin: 'left',
+    ops.InnerJoin: 'inner',
+    ops.OuterJoin: 'outer',
+}
+
+
+@execute_node.register(ops.Join, pd.DataFrame, pd.DataFrame)
+def execute_join(op, left, right, scope=None):
+    try:
+        how = _JOIN_TYPES[type(op)]
+    except KeyError:
+        raise NotImplementedError('{} not supported'.format(type(op).__name__))
+
+    left_on = []
+    right_on = []
+
+    for predicate in op.predicates:
+        pred_op = predicate.op()
+        if not isinstance(pred_op, ops.Equals):
+            raise TypeError(
+                'Only equality join predicates supported with pandas'
+            )
+        left_key = pred_op.left._name
+        right_key = pred_op.right._name
+        left_on.append(left_key)
+        right_on.append(right_key)
+
+    return pd.merge(left, right, how=how, left_on=left_on, right_on=right_on)
+
+
+_BINARY_OPERATIONS = {
+    ops.Greater: operator.gt,
+    ops.Less: operator.lt,
+    ops.LessEqual: operator.le,
+    ops.GreaterEqual: operator.ge,
+    ops.Equals: operator.eq,
+    ops.NotEquals: operator.ne,
+
+    ops.And: operator.and_,
+    ops.Or: operator.or_,
+
+    ops.Add: operator.add,
+    ops.Subtract: operator.sub,
+    ops.Multiply: operator.mul,
+    ops.Divide: operator.truediv,
+    ops.Power: operator.pow,
+}
+
+
+@execute_node.register(
+    ops.BinaryOp, pd.Series, (pd.Series, numbers.Real) + six.string_types
+)
+def execute_binary_operation_series(op, left, right, scope=None):
+    return _BINARY_OPERATIONS[type(op)](left, right)
+
+
+def hashable(obj):
+    try:
+        hash(obj)
+    except TypeError:
+        return False
+    else:
+        return True
+
+
+def find_data(expr):
+    stack = [expr]
+    seen = set()
+
+    while stack:
+        e = stack.pop()
+        node = e.op()
+
+        if node not in seen:
+            seen.add(node)
+
+            if hasattr(node, 'source'):
+                yield {e: node.source.dictionary[node.name]}
```

Comment:
We should build the dictionary here rather than yielding them and merging later.